# Homework

In [3]:
#Handy imports from class

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os
import plotly.express as px
from scipy import signal

df = pd.read_csv('temperature.csv')

### 1. Write a function that efficiently converts every column to a different temperature scale

In [85]:
df

datetime   Vancouver    Portland  San Francisco     Seattle  \
0     2012-10-01 12:00:00         NaN         NaN            NaN         NaN   
1     2012-10-01 13:00:00  284.630000  282.080000     289.480000  281.800000   
2     2012-10-01 14:00:00  284.629041  282.083252     289.474993  281.797217   
3     2012-10-01 15:00:00  284.626998  282.091866     289.460618  281.789833   
4     2012-10-01 16:00:00  284.624955  282.100481     289.446243  281.782449   
...                   ...         ...         ...            ...         ...   
45248 2017-11-29 20:00:00         NaN  282.000000            NaN  280.820000   
45249 2017-11-29 21:00:00         NaN  282.890000            NaN  281.650000   
45250 2017-11-29 22:00:00         NaN  283.390000            NaN  282.750000   
45251 2017-11-29 23:00:00         NaN  283.020000            NaN  282.960000   
45252 2017-11-30 00:00:00         NaN  282.280000            NaN  283.040000   

       Los Angeles   San Diego   Las Vegas     Phoenix  Albuquerque  ...  \
0              NaN         NaN         NaN         NaN          NaN  ...   
1       291.870000  291.530000  293.410000  296.600000   285.120000  ...   
2       291.868186  291.533501  293.403141  296.608509   285.154558  ...   
3       291.862844  291.543355  293.392177  296.631487   285.233952  ...   
4       291.857503  291.553209  293.381213  296.654466   285.313345  ...   
...            ...         ...         ...         ...          ...  ...   
45248   293.550000  292.150000  289.540000  294.710000   285.720000  ...   
45249   295.680000  292.740000  290.610000  295.590000   286.450000  ...   
45250   295.960000  292.580000  291.340000  296.250000   286.440000  ...   
45251   295.650000  292.610000  292.150000  297.150000   286.140000  ...   
45252   294.930000  291.400000  291.640000  297.150000   284.700000  ...   

        Beersheba  Tel Aviv District       Eilat  Haifa  Nahariyya  Jerusalem  \
0             NaN                NaN  309.100000    NaN        NaN        NaN   
1      307.590000         305.470000  310.580000  304.4      304.4      303.5   
2      307.590000         304.310000  310.495769  304.4      304.4      303.5   
3      307.391513         304.281841  310.411538  304.4      304.4      303.5   
4      307.145200         304.238015  310.327308  304.4      304.4      303.5   
...           ...                ...         ...    ...        ...        ...   
45248         NaN                NaN         NaN    NaN        NaN        NaN   
45249         NaN                NaN         NaN    NaN        NaN        NaN   
45250         NaN                NaN         NaN    NaN        NaN        NaN   
45251         NaN                NaN         NaN    NaN        NaN        NaN   
45252         NaN                NaN         NaN    NaN        NaN        NaN   

       year  month  Seasons        date  
0        10     10     fall  2012-10-01  
1        10     10     fall  2012-10-01  
2        10     10     fall  2012-10-01  
3        10     10     fall  2012-10-01  
4        10     10     fall  2012-10-01  
...     ...    ...      ...         ...  
45248    11     11     fall  2017-11-29  
45249    11     11     fall  2017-11-29  
45250    11     11     fall  2017-11-29  
45251    11     11     fall  2017-11-29  
45252    11     11     fall  2017-11-30  

[45253 rows x 41 columns]

In [14]:
#this converts columns from Kelvin to Celcius
def convert(m):
    for i in m:
        m = m - 273.15
        return m

convert(df["Boston"])

0              NaN
1        14.020000
2        14.036092
3        14.081672
4        14.127251
           ...    
45248    14.930000
45249    12.870000
45250    10.790000
45251     9.020000
45252     7.500000
Name: Boston, Length: 45253, dtype: float64

### 2. Write a function that efficiently fills in missing values in every column, except for trailing NANs.  

That is: if there's temperature data on either side of a timestamp, it should interpolate through it. But if data stopped being collected, your function should leave nans at the end

In [32]:
#df.ffill().where(df.bfill().notnull()) # was just experimenting with different ways to do it

In [33]:
df['datetime'] = pd.to_datetime(df['datetime'])

x = df['datetime']
knowns = df.loc[~df['Vancouver'].isnull(), ['datetime','Vancouver']]
xp = knowns['datetime']
fp = np.array(knowns['Vancouver'])

test = pd.DataFrame({'datetime':x, 'Vancouver': np.interp(x, xp, fp)})

In [28]:
test

datetime   Vancouver
0     2012-10-01 12:00:00  284.630000
1     2012-10-01 13:00:00  284.630000
2     2012-10-01 14:00:00  284.629041
3     2012-10-01 15:00:00  284.626998
4     2012-10-01 16:00:00  284.624955
...                   ...         ...
45248 2017-11-29 20:00:00  288.010000
45249 2017-11-29 21:00:00  288.010000
45250 2017-11-29 22:00:00  288.010000
45251 2017-11-29 23:00:00  288.010000
45252 2017-11-30 00:00:00  288.010000

[45253 rows x 2 columns]

In [34]:
df.interpolate(method ='linear', limit_direction ='forward')

datetime   Vancouver    Portland  San Francisco     Seattle  \
0     2012-10-01 12:00:00         NaN         NaN            NaN         NaN   
1     2012-10-01 13:00:00  284.630000  282.080000     289.480000  281.800000   
2     2012-10-01 14:00:00  284.629041  282.083252     289.474993  281.797217   
3     2012-10-01 15:00:00  284.626998  282.091866     289.460618  281.789833   
4     2012-10-01 16:00:00  284.624955  282.100481     289.446243  281.782449   
...                   ...         ...         ...            ...         ...   
45248 2017-11-29 20:00:00  288.010000  282.000000     302.070000  280.820000   
45249 2017-11-29 21:00:00  288.010000  282.890000     302.070000  281.650000   
45250 2017-11-29 22:00:00  288.010000  283.390000     302.070000  282.750000   
45251 2017-11-29 23:00:00  288.010000  283.020000     302.070000  282.960000   
45252 2017-11-30 00:00:00  288.010000  282.280000     302.070000  283.040000   

       Los Angeles   San Diego   Las Vegas     Phoenix  Albuquerque  ...  \
0              NaN         NaN         NaN         NaN          NaN  ...   
1       291.870000  291.530000  293.410000  296.600000   285.120000  ...   
2       291.868186  291.533501  293.403141  296.608509   285.154558  ...   
3       291.862844  291.543355  293.392177  296.631487   285.233952  ...   
4       291.857503  291.553209  293.381213  296.654466   285.313345  ...   
...            ...         ...         ...         ...          ...  ...   
45248   293.550000  292.150000  289.540000  294.710000   285.720000  ...   
45249   295.680000  292.740000  290.610000  295.590000   286.450000  ...   
45250   295.960000  292.580000  291.340000  296.250000   286.440000  ...   
45251   295.650000  292.610000  292.150000  297.150000   286.140000  ...   
45252   294.930000  291.400000  291.640000  297.150000   284.700000  ...   

       Philadelphia    New York    Montreal      Boston   Beersheba  \
0               NaN         NaN         NaN         NaN         NaN   
1        285.630000  288.220000  285.830000  287.170000  307.590000   
2        285.663208  288.247676  285.834650  287.186092  307.590000   
3        285.756824  288.326940  285.847790  287.231672  307.391513   
4        285.850440  288.406203  285.860929  287.277251  307.145200   
...             ...         ...         ...         ...         ...   
45248    290.240000  284.980000  275.130000  288.080000  285.944000   
45249    289.240000  284.980000  274.130000  286.020000  285.944000   
45250    286.780000  284.980000  273.480000  283.940000  285.944000   
45251    284.570000  284.980000  272.480000  282.170000  285.944000   
45252    283.420000  284.980000  271.800000  280.650000  285.944000   

       Tel Aviv District       Eilat    Haifa  Nahariyya  Jerusalem  
0                    NaN  309.100000      NaN        NaN        NaN  
1             305.470000  310.580000  304.400    304.400     303.50  
2             304.310000  310.495769  304.400    304.400     303.50  
3             304.281841  310.411538  304.400    304.400     303.50  
4             304.238015  310.327308  304.400    304.400     303.50  
...                  ...         ...      ...        ...        ...  
45248         294.150000  291.019000  299.369    299.369     294.15  
45249         294.150000  291.019000  299.369    299.369     294.15  
45250         294.150000  291.019000  299.369    299.369     294.15  
45251         294.150000  291.019000  299.369    299.369     294.15  
45252         294.150000  291.019000  299.369    299.369     294.15  

[45253 rows x 37 columns]

### 3. Find cities A and B where city A is warmer than city B in the summer, but also colder than city B in the winter, on average

In [39]:
df['month'] = pd.DatetimeIndex(df['datetime']).month
df['month']

0        10
1        10
2        10
3        10
4        10
         ..
45248    11
45249    11
45250    11
45251    11
45252    11
Name: month, Length: 45253, dtype: int64

In [74]:
season_bins = [1,3,6,9,12] #creating bins to group months by season to figure our the average temparatures for winter and summer
season_labels = ["winter", "spring", "summer", "fall"]

df["Seasons"] = pd.cut(df["month"], season_bins, labels=season_labels )

grouped_seasons = df.groupby("Seasons")

SeasonAverage = {"vanc_temp":grouped_seasons["Vancouver"].mean(),
                "san_diego_temp":grouped_seasons["San Diego"].mean(),
                 "las_vegas_temp":grouped_seasons["Las Vegas"].mean(),
                "portland_temp":grouped_seasons["Portland"].mean(),
                "los_ang_temp":grouped_seasons["Los Angeles"].mean()}

pd.DataFrame(SeasonAverage)

vanc_temp  san_diego_temp  las_vegas_temp  portland_temp  \
Seasons                                                              
winter   279.098436      287.374702      286.394675     280.488674   
spring   286.358575      290.551596      297.196965     287.349380   
summer   290.963652      295.002449      303.086515     292.491610   
fall     280.030428      288.722986      285.855432     281.320055   

         los_ang_temp  
Seasons                
winter     287.729062  
spring     291.541748  
summer     296.235125  
fall       288.817639

City A is Las Vegas and city B is San Diego. Las Vegas is warmer than San Diego in summer(303 degrees) and colder in winter(286). These 2 cities satisfy the assignment requirement.

### 4. For the two cities you found in for problem 4 (there are many possible answers), clean up the data via aggregation or rolling averages and plot a clear visual

In [76]:
df['date']=df['datetime'].dt.date

In [83]:
px.line(df.groupby('date').agg({'San Diego':'mean'}).reset_index(), x='date', y='San Diego')

In [84]:
px.line(df.groupby('date').agg({'Las Vegas':'mean'}).reset_index(), x='date', y='Las Vegas')